internal/pbutil/shallow_equal.go

```diff
@@ -0,0 +1,54 @@
+package pbutil
+
+import (
+	"errors"
+	"reflect"
+)
+
+// ShallowEqual compares two things in a shallow (not deep like reflect.DeepEqual) way.
+// if they're pointers, they're dereferenced first before proceeding.
+// if they're structs, it does Value.Equal() for all visible fields whose types are .Comparable(), and ignores any other fields.
+// otherwise, if they're some other type that is .Comparable(), it just does Value.Equal() and hopes for the best.
+// if none of the above apply, it returns false.
+// superceding any of the above, if an error occurs, false will be returned along with the error.
+// note that this method is different from reflect.Equal in that it does not care if the types are the same, only the kinds.
+func ShallowEqual(a, b any) (bool, error) {
+	aVal, bVal := reflect.Indirect(reflect.ValueOf(a)), reflect.Indirect(reflect.ValueOf(b))
+	if aVal.Kind() == reflect.Struct && bVal.Kind() == reflect.Struct {
+		aFields, bFields := reflect.VisibleFields(aVal.Type()), reflect.VisibleFields(bVal.Type())
+		aComparableFields, bComparableFields := Filter(aFields, fieldTypeIsComparable), Filter(bFields, fieldTypeIsComparable)
+		if len(aComparableFields) != len(bComparableFields) {
+			return false, nil
+		}
+		for i, aField := range aComparableFields {
+			aFieldVal, aErr := aVal.FieldByIndexErr(aField.Index)
+			bFieldVal, bErr := bVal.FieldByIndexErr(bComparableFields[i].Index)
```

Comment:
Since A and B are `any`, they can be 2 different types, correct? `reflect.VisibleFields` doesn't sort fields and return in `the same order found in the struct` https://pkg.go.dev/reflect#VisibleFields This means that even if 2 types have the same exported fields `bComparableFields[i]` can point to the different field than `aComparableFields[i]`. Did I get the logic of this method correctly? Should we sort fields in `Filter` function?

Reply:
> Since A and B are `any`, they can be 2 different types, correct?

Yes, I wanted to support polymorphism to some degree.

> `reflect.VisibleFields` doesn't sort fields and return in `the same order found in the struct`

Unless I'm grossly misunderstanding something, the docs you link to say otherwise (emphasis mine):

> VisibleFields returns all the visible fields in t, which must be a struct type. A field is defined as visible if it's accessible directly with a FieldByName call. The returned fields include fields inside anonymous struct members and unexported fields. **They follow the same order found in the struct**, with anonymous fields followed immediately by their promoted fields. 

> This means that even if 2 types have the same exported fields `bComparableFields[i]` can point to the different field than `aComparableFields[i]`. Did I get the logic of this method correctly?

Yes, that's correct, I was not sure whether I should prioritize field order or field names. On one hand, two structs that have identical contents but call them something differently feel like they should be considered equal, but on the other hand, so do two structs that use identical names for their fields but list them in different orders, where each respective field by name has the same value. I could not decide on which was more important, so I just submitted the first one I wrote and got to work; perhaps I should try both approaches and if either say the structs are equal, then this function should return `true`. What do you think?

> Should we sort fields in `Filter` function?

No, I think that would be a pain. There is a [`FieldByName`](https://pkg.go.dev/reflect#Value.FieldByName) function I could use on `aVal` and `bVal`, either in this loop or in a second loop that only happens if this loop would return a `false` result.

Reply:
Let's add the comment in this case that
```
{
  a = "a"
  b = "b"
}
```
is not equal to
```
{
  b = "b"
  a = "a"
}
```

Reply:
Yeah I could at the very least document that this function cares about struct member order not names as it is written. But if you would like me to have it check order or names and accept equality if either approach finds a match, then just let me know and I'll be happy to do it.

Reply:
Since I didn't get any word from you, I decided to have it check both. Please let me know what you think.